frontend/catalyst/utils/jax_extras.py

```diff
@@ -477,3 +576,372 @@ def make_jaxpr_f(*args, **kwargs):
+        out_tracers.append(maker(aval))
+    return out_tracers
+
+
+TracerLike = TypeVar("TracerLike")
+
+
+def infer_input_type_unshared(inputs: List[TracerLike]) -> InputType:
+    """Infer the input type of a function having `inputs` Jax tracers."""
+
+    def _is_tracer_like(x):
+        return hasattr(x, "aval")
+
+    expl_ins = inputs
+    impl_avals = []
+    expl_avals = []
+    for o in expl_ins:
+        assert _is_tracer_like(o), (o,)
+        if isinstance(o.aval, DShapedArray):
+            shape2 = []
+            for d in o.aval.shape:
+                if _is_tracer_like(d):
+                    shape2.append(DBIdx(len(impl_avals)))
+                    impl_avals.append(d.aval)
+                else:
+                    shape2.append(d)
+            expl_avals.append(o.aval.update(shape=tuple(shape2)))
+        else:
+            expl_avals.append(o.aval)
+    return (*[(i, False) for i in impl_avals], *[(i, True) for i in expl_avals])
+
+
+@dataclass
+class ExpansionStrategy:
+    """Describe the settings affecting the Jax dyanmic API tracing of the nested programs.
+    Args:
+        axes_specs: Axes specification used to convert static dimensions to the dynamic ones
+        input_unshare_variables: Treat each dynamic dimension as a distinct dimension, even if some
+                                 of them are described by the same dimension variables.
+        output_include_indbidx_vars: Include variables corresponding to the InDBIdx references in
+                                     the result
+        output_force_arg0_outdbidx: Force all references pointing to the first arguments to be
+                                    OutDBIdx. This is typically required to avoid input references
+                                    to the loop iteration variable.
+    """
+
+    axes_specs: Sequence[AbstractedAxesSpec] | None
+    input_unshare_variables: bool
+    output_include_indbidx_vars: bool
+    output_force_arg0_outdbidx: bool
+
+
+def while_loop_expansion_strategy(preserve_dimensions=False):
+    """Arguments and results expansion strategy for while-loops."""
+    return ExpansionStrategy(None, not preserve_dimensions, True, False)
+
+
+def for_loop_expansion_strategy(preserve_dimensions=False):
+    """Arguments and results expansion strategy for for-loops."""
+    return ExpansionStrategy(None, not preserve_dimensions, True, True)
+
+
+def cond_expansion_strategy():
+    """Arguments and results expansion strategy for conditionals."""
+    return ExpansionStrategy(None, False, True, False)
+
+
+def infer_output_type(
+    constants: List[TracerLike],
+    expanded_inputs: List[TracerLike],
+    outputs: List[TracerLike],
+    expansion_strategy: ExpansionStrategy,
+    num_implicit_inputs: int | None = None,
+) -> Tuple[List[TracerLike], OutputType]:
+    """Deduce the Jax ``out_type`` given input and ouputs abstract entities. By abstract entities
+    we mean either Jax tracers or Jaxpr variables."""
```

Comment:
What exactly is a Jaxpr variable (what is its Python class?) as opposed to a Jax tracer?

Reply:
Both are distinct explicitly defined entities. Here is the [Jaxpr variable](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L278) and here is the [Jax tracer](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L1522).
I see it as follows. Jax understands programs in two languages: Jaxpr (their own internal language) and Python.  Jax need to operate on lists of arguments and results somehow. So they use lists of tracers (in the context of tracing) or lists of Jaxpr variables interchangebly. See also this [Jax comment](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L1677).